### ISSUE TITLE
Fieldsets that contain a checkbox grouping does not read the legend at times

### ISSUE THREAD

**Issue by nvaccessAuto** [nvaccess/nvda#3321]
**Reported by nshah9 on 2013-07-01 19:41**
I have 2 pages with checkbox groupings.

On one of the pages, when the grouping receives focus, the screen reader reads the legend as expected. On this page, the checkbox grouping is surrounded by a series of other inputs.

However, on the second page, when the grouping receives focus, the screen reader does not read the legend, which I think is an NVDA bug. If the grouping receives focus from the bottom, i.e (Shift+Tab), the screen reader reads the legend correctly. On this page, the checkbox grouping has a link before it and other inputs after it.

**Comment by afiorenza** [comment]
This bugs also happens in firefox with nvda, but seem to fails randomly. Any news about this?

**Comment by weboverhauls** [comment]
Please fix. 
I can also reproduce with NVDA/Firefox.

**Comment by RichCaloggero** [comment]
Perhaps the rule could be something like:
- if last keyboard event was generated via arrow key, assume browse mode and do not announce group label
- otherwise always announce

This would seem to take in to account the case where focus was shifted via javascript into the group (in which case label should be announced), or when tabbing (also should be announced). If last key pressed was arrow key then assume the user does not need the group label announced.

**Comment by HerinHentry** [comment]
Yes. This issue occurs in Firefox & IE. Thanks

**Comment by jcsteh** [comment]
P2 because this causes users to miss important information they would normally receive when tabbing.

**Comment by JLastort** [comment]
I found a workaround. You can indicate to NVDA that it should go into forms mode by wrapping the fieldset in a div.

<div role="form" aria-label="Type of Service">
<fieldset>
....
</div>

**Comment by grahamarmfield** [comment]
> 
> 
> P2 because this causes users to miss important information they would normally receive when tabbing.

This for me makes the issue quite important. The legend not being read out can cause confusion to users, and potentially lead to incorrect choices.

Is there any likelihood of a solution for this?

**Comment by jcsteh** [comment]
This issue has already been fixed. If you can still reproduce it, please provide a test case and steps to reproduce. Thanks.

**Comment by StommePoes** [comment]
I still don't hear the legend: http://www.stommepoes.nl/work/k-12/fieldsets.html (test #3)

We just figured we had to hope users would wander around and manage to get into forms mode. :(

**Comment by AlmeroSteyn** [comment]
Applying the pattern suggested above by @StommePoes to radio buttons and checkboxes the result is this:

http://debonair-scissors.surge.sh

It appears that NVDA goes out of forms mode when encountering checkboxes.

**Comment by RichCaloggero** [comment]
>It appears that NVDA goes out of forms mode when encountering checkboxes.

This makes sense since the only reason for forms mode (non-virtual or real mode) is to turn off all screen reader keyboard processing so the browser can use all the keys.  This is necessary, for instance, in a text input field where you want arrows to move the caret, delete / backspace to delete characters, etc.  A checkbox and button have no such requirements, so the screen reader does not turn off virtual mode when encountering these. Radio buttons, however, do require the arrow keys to move among them and check the currently selected one.  Note that radio buttons, in order to be seen by the browser as a single group, need to all be associated with the same `name` attribute:

```
<!-- note that headings can now be children of legend -- very convenient and some situations -->
<!-- see http://validator.w3.org/nu/ -->
<fieldset><legend><h2>radio buttons</h2></legend>
<label>radio button 1 <input type="radio" name="group"></label>
<label>radio button 1 <input type="radio" name="group"></label>
</fieldset>
```

As per the bug about which this issue is concerned: seems fixed to me with latest NVDA!

**Comment by RichCaloggero** [comment]
>It appears that NVDA goes out of forms mode when encountering checkboxes.

This makes sense since the only reason for forms mode (non-virtual or real mode) is to turn off all screen reader keyboard processing so the browser can use all the keys.  This is necessary, for instance, in a text input field where you want arrows to move the caret, delete / backspace to delete characters, etc.  A checkbox and button have no such requirements, so the screen reader does not turn off virtual mode when encountering these. Radio buttons, however, do require the arrow keys to move among them and check the currently selected one.  Note that radio buttons, in order to be seen by the browser as a single group, need to all be associated with the same `name` attribute:

```
<!-- note that headings can now be children of legend -- very convenient and some situations -->
<!-- see http://validator.w3.org/nu/ -->
<fieldset><legend><h2>radio buttons</h2></legend>
<label>radio button 1 <input type="radio" name="group"></label>
<label>radio button 1 <input type="radio" name="group"></label>
</fieldset>
```

As per the bug about which this issue is concerned: seems fixed to me with latest NVDA!

**Comment by StommePoes** [comment]
Fixed as in, you hear legends now when coming from an error link? Because errors as links going to the errant fields is the current recommendation for forms, and here it breaks (2018.3 with at least one update though I think I got the other one after testing).
